src/ir/datalog/input_relation_fact.h

```diff
@@ -0,0 +1,56 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+#ifndef SRC_IR_DATALOG_INPUT_RELATION_FACT_H_
+#define SRC_IR_DATALOG_INPUT_RELATION_FACT_H_
+
+#include <utility>
+
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_join.h"
+#include "src/ir/datalog/value.h"
+
+namespace raksha::ir::datalog {
+
+// A fact that is built to be fed into an input relation for Souffle.
+// Although the `Rule` class can also express a `Fact` by having no RHS, this
+// is explicitly to prepare input to a Souffle program. You can think of the
+// difference between this `Fact` and a `Rule` fact as a difference between
+// data and code.
+template <class... RelationParameterTypes>
+class InputRelationFact {
+ public:
+  InputRelationFact(RelationParameterTypes &&...args)
+      : relation_arguments_(std::forward<RelationParameterTypes>(args)...) {}
+
+  virtual absl::string_view GetRelationName() const = 0;
```

Comment:
Add a string template argument for the relation name and replace this virtual function with a regular one? That will eliminate a lot of boiler place you see in the tests and can define classes more directly:

```C++
using IsAccessPathFact = InputRelationFact<"IsAccessPath", Symbol>;
```

Reply:
The approach you suggest is the one I tried first. But I ran into subtle problems. The issue is that only pointers to objects without linkage are allowed as template arguments. That means that, for a derived class with a superclass with a string template argument defined in a header file, the pointer and thus the superclass will be different for every TU that the header gets linked into. That problem is detected by the `-Wsubobject-linkage` warning in GCC and prevented me from getting into very confusing territory.

I think that the way forward for this (and some of the other patterns) may actually be to create macros to hide the low-entropy parts of the definition. You could imagine one for defining `InputRelationFact`s like this:

```C++
#define INPUT_RELATION_FACT_CLASS_DECL(factRelationName) \
  class InputRelationFact_ ## factRelationName
  
#define INPUT_RELATION_FACT_DEFN(factRelationName) \
  INPUT_RELATION_FACT_CLASS_DECL(factRelationName) : public InputRelationFact { \
    using InputRelationFact::InputRelationFact;
    static absl::string_view relation_name() { return #factRelationName ; } \
    absl::string_view GetRelationName() const override { return relation_name(); } \
  }
```

Note that by using the relation name to both generate the class name and produce the string output by `GetRelationName`, we prevent two classes linked into the same translation unit from having different class names but being for the same relation.